File: src/java/MatrixInversion.java
import java.io.IOException;
import java.util.Arrays;

public class MatrixInversion {

    public static int hasArgument(String[] args, String arg) {
        for (int i = 0; i < args.length; i++) {
            if (args[i].equals(arg))
                return i;
        }
        return -1;
    }

    public static Matrix[] getLUDecomposition(float[][] a) {
        int N = a.length;
        Matrix l = new Matrix(N);
        Matrix u = new Matrix(N);

        // By default everything in the array is 0.0
        float[][] alpha = l.m;
        float[][] beta = u.m;

        for (int i = 0; i < N; i++) {
            alpha[i][i] = 1;
        }

        for (int j = 0; j < N; j++) {
            for (int i = 0; i <= j; i++) {
                float sum = 0.0f;
                for (int k = 0; k < i; k++) {
                    sum += alpha[i][k] * beta[j][k];
                }
                beta[j][i] = a[i][j] - sum;
            }
            for (int i = j + 1; i < N; i++) {
                float sum = 0;
                var alpha_p = alpha[i];
                var beta_p = beta[j];
                for (int k = 0; k < j; k++) {
                    sum += alpha_p[k] * beta_p[k];
                }
                alpha[i][j] = (1 / beta[j][j]) * (a[i][j] - sum);
            }

        }

        transposeMatrix(beta);

        return new Matrix[] { l, u };
    }

    public static void transposeMatrix(float[][] m) {
        for (int i = 0; i < m.length; i++) {
            for (int j = 0; j < i; j++) {
                float tmp = m[i][j];
                m[i][j] = m[j][i];
                m[j][i] = tmp;
            }
        }
    }

    public static float[] findY(float[][] alpha, float[] b) {
        int N = alpha.length;
        float[] y = new float[N];

        y[0] = b[0] / alpha[0][0];

        for (int i = 1; i < N; i++) {
            float sum = 0.0f;
            for (int j = 0; j < i; j++) {
                sum += alpha[i][j] * y[j];
            }
            y[i] = (b[i] - sum) / alpha[i][i];
        }

        return y;
    }

    public static float[] findX(float[][] beta, float[] y) {
        int N = beta.length;
        float[] x = new float[N];

        x[N - 1] = y[N - 1] / beta[N - 1][N - 1];

        for (int i = N - 2; i >= 0; i--) {

            float sum = 0.0f;

            for (int j = i; j < N; j++) {
                sum += beta[i][j] * x[j];
            }

            x[i] = (y[i] - sum) / beta[i][i];
        }

        return x;

    }

    public static float[][] getIdentityMatrix(int N) {
        float[][] E = new float[N][N];

        for (int i = 0; i < N; i++) {
            E[i][i] = 1;
        }
        return E;
    }

    public static float[][] inverse(float[][] alpha, float[][] beta) {
        int N = alpha.length;

        float[][] E = getIdentityMatrix(N);

        float[][] X = new float[N][N];

        for (int i = 0; i < N; i++) {
            float[] y = findY(alpha, E[i]);
            System.out.print("[");
            for (int j = 0; j < N; j++)
                System.out.print(y[j] + ", ");
            System.out.println("]");
            X[i] = findX(beta, y);
        }

        transposeMatrix(X);

        return X;

    }

    public static void main(String[] args) throws IOException {
        Matrix matrix = new Matrix(args[args.length - 2]);

        long before = System.nanoTime();
        Matrix[] matrices = getLUDecomposition(matrix.m);
        Matrix l = matrices[0];
        Matrix u = matrices[1];
        float[][] X = inverse(l.m, u.m);
        long after = System.nanoTime();
        long runtime = after - before;

        if (Arrays.stream(args).anyMatch("--time"::equals)) {
            System.out.println(runtime);
        }
        if (Arrays.stream(args).anyMatch("--printresult"::equals)) {
            Matrix.printMatrix(l.m);
            Matrix.printMatrix(u.m);
            Matrix.printMatrix(X);
        }
        // int outputResult = hasArgument(args, "--outputresult");
        // if (outputResult > -1) {
        // Matrix.outputMatrix(result, args[outputResult + 1]);
        // }
    }
}

File: src/c/matrix_inversion.c
#include <stdio.h>
#include "../utilities/matrix.h"
#include "../utilities/utils.h"

#define IDX(i,j,size) (((i) * (size)) + (j))
void inverse_matrix(float* res, float* a, int size, int small_size);


int main(int argc, char** argv) {
    char* path1 = argv[argc - 2];
    char* path2 = argv[argc - 1];
    int displayRuntime = contains_argument(argc, argv, "--time");
    int print_to_file = contains_argument(argc, argv, "--outputresult");
    int print_result = contains_argument(argc, argv, "--printresult");
    int load_time = contains_argument(argc, argv, "--loadtime");

    unsigned long before_load = get_time_nanoseconds();
    Matrices* ma = load_matrices(path1, path2);
    unsigned long after_load = get_time_nanoseconds();

    int total_size = ma->total_size;
    //printMatrix(matrix1, size);
    //printMatrix(matrix2, size);
    unsigned long before = get_time_nanoseconds();
    inverse_matrix(ma->result, ma->m1, total_size, ma->size);
    unsigned long after = get_time_nanoseconds();
    unsigned long runtime = after - before;
    //printMatrix(result, size);
    if(load_time) 
        printf("load time: %lu\n", after_load - before_load);

    if(displayRuntime) 
        printf("%lu\n", runtime);

    if (print_result) {
        print_matrix(ma->result, ma->size);
    }

    if(print_to_file){
        char* path = argv[print_to_file+1];
        output_matrix(ma, path);
    }
    free_matrices(ma);
    return 0;
}

void transpose_matrix(float* m, int N){
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < i; j++) {
                float tmp = m[IDX(i,j,N)];
                m[IDX(i,j,N)] = m[IDX(j,i,N)];
                m[IDX(j,i,N)] = tmp;
            }
        }
}

void LU_decompose(float* alpha, float* beta, float* a, int N){
        for (int i = 0; i < N; i++) {
            alpha[IDX(i, i, N)] = 1;
        }

        for (int j = 0; j < N; j++) {
            for (int i = 0; i <= j; i++) {
                float sum = 0.0f;
                for (int k = 0; k < i; k++) {
                    sum += alpha[IDX(i,k,N)] * beta[IDX(j,k,N)];
                }
                beta[IDX(j,i,N)] = a[IDX(i,j,N)] - sum;
            }
            for (int i = j + 1; i < N; i++) {
                float sum = 0;
                float* alpha_p = &alpha[IDX(i,0, N)];
                float* beta_p = &beta[IDX(j,0,N)];
                for (int k = 0; k < j; k++) {
                    sum += alpha_p[k] * beta_p[k];
                }
                alpha[IDX(i,j,N)] = (1 / beta[IDX(j,j,N)]) * (a[IDX(i,j,N)] - sum);
            }
        
        }

        transpose_matrix(beta, N);
}

void findY(float* y, float* alpha, float* b, int N){

    y[0] = b[0] / alpha[0];

    for (int i = 1; i < N; i++) {
        float sum = 0.0f;
        for (int j = 0; j < i; j++) {
            sum += alpha[IDX(i,j,N)] * y[j];
        }
        y[i] = (b[i] - sum) / alpha[IDX(i,i,N)];
    }
}


void findX(float* x, float* beta, float* y, int N){

    x[N - 1] = y[N - 1] / beta[IDX(N - 1,N - 1,N)];

    for (int i = N - 2; i >= 0; i--) {

        float sum = 0.0f;

        for (int j = i+1; j < N; j++) {
            sum += beta[IDX(i,j,N)] * x[j];
        }
        x[i] = (y[i] - sum) / beta[IDX(i,i,N)];
    }
}

void populate_identity_matrix(float* E, int N){
        for (int i = 0; i < N; i++) {
            E[IDX(i,i,N)] = 1;
        }
}

void print_float_pointer(char* name, float* p, int N){
        printf("%s:\n", name);
        for(int z = 0; z < N; z++) {
            printf("%f, ", p[z]);
        }
        printf("\n");
}

void inverse_matrix(float* res, float* a, int size, int small_size) {
    float* alpha = malloc(size*sizeof(float));
    float* beta = malloc(size*sizeof(float));
    float* E = calloc(size, sizeof(float));
    LU_decompose(alpha, beta, a, small_size);

    populate_identity_matrix(E, small_size);


    float* y = malloc(small_size * sizeof(float));
    float* x = malloc(small_size * sizeof(float));
    for (int i = 0; i < small_size; i++) {
        findY(y, alpha, &E[IDX(i, 0, small_size)], small_size);
        findX(x, beta, y, small_size);
        for (int j = 0; j < small_size; j++) {
            res[IDX(i, j, small_size)] = x[j];
        }
    }

    transpose_matrix(res, small_size);

    free(alpha);
    free(beta);
    free(E);
    free(y);
    free(x);

}
